04_setup_ironic.sh

```diff
@@ -49,18 +49,32 @@ for IMAGE_VAR in $(env | grep "_LOCAL_IMAGE=" | grep -o "^[^=]*") ; do
         REPOPATH=~/${IMAGE##*/}
         # Clone to ~ if not there already
         [ -e "$REPOPATH" ] || git clone $IMAGE $REPOPATH
+
+        # If we set extra packages, we add them to the main pkgs base file
+        MAIN_PKGS_FILE_NAME=${IMAGE_VAR/_LOCAL_IMAGE}_MAIN_PKGS
+        MAIN_PKGS_FILE=${!MAIN_PKGS_FILE_NAME:-"main-packages-list.txt"}
+        EXTRA_PKGS_FILE_NAME=${IMAGE_VAR/_LOCAL_IMAGE}_EXTRA_PACKAGES
+        EXTRA_PKGS_FILE=${!EXTRA_PKGS_FILE_NAME:-}
+        if [[ -n $EXTRA_PKGS_FILE ]]; then
+            # add newline at the end of file if not present
+            sed -i -e '$a\' "${REPOPATH}/${MAIN_PKGS_FILE}"
+            cat "$EXTRA_PKGS_FILE" >> "${REPOPATH}/${MAIN_PKGS_FILE}"
+        fi
+
```

Comment:
Hi Riccardo,
not sure if I missed something, but if it's for local testing (and thus there's already a repo checked out), why can't you modify directly the pkg base file? In any case here you have to define an extra file and store it somewhere. Can you please clarify thanks?

Reply:
if you build an image locally, the image repository needs to be synced the first time, and that's done automatically by the script
you won't have any file to change the first time you run dev-scripts with locally built images, so without this you will need to manually clone the repo, modify the packages file, and then run the script
another option I considered is to modify the images  themselves to accept an "extra package" file directly as argument, that would simplify things, but it requires more work (need to change 3 images first, then dev-scripts).
I will probably do it anyway, depending on which direction we take to build the final images, but for now this is the fastest solution IMHO

Reply:
Am I missing something or #1140 could cover also such case, without the need of passing an extra file (but instead by directly modifying the Dockerfile)?

Reply:
#1084 is needed just to import and test a PR for one or more images
it doesn't involve changing the Dockerfile

Reply:
As far I as understood this PR it's useful to test manually the changes to the extra packages _without_ checking out and modifying explicitly the target repo, and that's ok.
 
Assuming that the tested changes are ok, a new PR on the target repo will be still required, containing the same new packages defined in the extra file. Under this point of view #1084 appears to be more CI-friendly, allowing to "revert" the logical workflow and to test directly the desired changes in dev-script (ie, a new PR `x` could be opened for `ironic-image` with some changes to https://github.com/openshift/ironic-image/blob/master/main-packages-list.txt, and then another PR in dev-script could be used to test PR `x`, thanks to the fact that extra config variables can now be injected directly).

Reply:
I think there's a bit of confusion here:
- this PR allows to install extra packages in the images during build time, no matter what they come from, and it's useful for testing packages before doing cross-tags, for example
- the #1084 allows to test PRs from the ironic container images no matter their content, but it can't test new packages if they're not cross-tagged first
they have 2 complete different scopes, this one is mainly for downstream testing, #1084 is more for upstream testing

Reply:
Right, but since technically speaking both of them are applied during the build time and could affect the same files as well, I'd suggest a couple of points:

1) Enrich a little bit the documentation and clarify the scope of usage of the config vars (in addition to the technical description, which is useful) 
2) For this PR, use an env var (instead of a file - or at least not just the file) to specify the extra packages to be tested: such approach could be more easily tested also via CI thanks to the usage of the `[DEVSCRIPTS_CONFIG](https://github.com/openshift/release/blob/0fb4497f32ee508cb8d692b2eae900611bbd889f/ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml#L10)` setting

Minor point: #1084 if defined at the same time could overwrite the context of _EXTRA_PACKAGES, probably it could be useful to add a check once this PR will be landed

Reply:
> Right, but since technically speaking both of them are applied during the build time and could affect the same files as well, I'd suggest a couple of points:
> 
> 1. Enrich a little bit the documentation and clarify the scope of usage of the config vars (in addition to the technical description, which is useful)

Will do

> 2. For this PR, use an env var (instead of a file - or at least not just the file) to specify the extra packages to be tested: such approach could be more easily tested also via CI thanks to the usage of the `[DEVSCRIPTS_CONFIG](https://github.com/openshift/release/blob/0fb4497f32ee508cb8d692b2eae900611bbd889f/ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml#L10)` setting

If I get it right, I disagree on this.
If we use an env var and we have to specify multiple packages, it will be kind of hard to set and read.
A file with one rpm per line is very easy to prepare, and it's very clear to read
Please consider that the logic in the images repos will slightly change, so this also will become a little different.
I'm actually tempted to go the other way around, change the images first and then here, it probably makes more sense.

> 
> Minor point: #1084 if defined at the same time could overwrite the context of _EXTRA_PACKAGES, probably it could be useful to add a check once this PR will be landed

Yes, I know, unfortunately.
This also makes me wonder if I should just handle the changes to the images first and then come back here, after #1084 is merged.

Reply:
@andfasano thanks for the discussion and the constructive feedback, it helped making me reconsider my approach
despite working correctly most of the times, in the end I feel like I need one more change on the images before moving forward

Reply:
Thanks, sounds good! :+1: